src/framework/utils.hpp

```diff
@@ -510,6 +510,15 @@ std::vector<T> vectorize_matrix(const matrix<T>& mat) {
   return vec;
 }
 
+template<class T>
+std::vector<T> vectorize_diagonal_matrix(const matrix<T>& mat) {
```

Comment:
This function actually just extracts the diagonal from a square matrix, so maybe we should call it `matrix_diagonal`